smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/RequestCompressionTraitValidator.java

```diff
@@ -0,0 +1,102 @@
+ */
+
+package software.amazon.smithy.model.validation.validators;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.traits.RequestCompressionTrait;
+import software.amazon.smithy.model.traits.RequiresLengthTrait;
+import software.amazon.smithy.model.traits.StreamingTrait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+
+/**
+ * Validates the requestCompression trait.
+ *
+ * - Validates that the operation input has no member with both streaming and
+ *   requiresLength traits applied
+ * - Validate encodings has at least one compression algorithm
+ * - Validate encodings are all supported compression algorithms
+ */
+public final class RequestCompressionTraitValidator extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        Set<OperationShape> shapes = model.getOperationShapesWithTrait(RequestCompressionTrait.class);
+        shapes.forEach(shape -> {
+            validateOperationInput(model, shape, events);
+            validateEncodings(model, shape, events);
+        });
+        return events;
+    }
+
+    private void validateOperationInput(Model model, OperationShape operationShape, List<ValidationEvent> events) {
+        // Validate operation input has no member with both streaming and requiresLength
+        // traits applied
+        if (operationShape.getInputShape().equals(ShapeId.from("smithy.api#Unit"))) {
+            events.add(error(operationShape,
+                    "The requestCompression trait can only be applied to operations which have an "
+                            + "input shape defined"));
+            return;
+        }
+        StructureShape inputShape = model.expectShape(operationShape.getInputShape(), StructureShape.class);
+        for (Map.Entry<String, MemberShape> entry : inputShape.getAllMembers().entrySet()) {
+            MemberShape memberShape = entry.getValue();
+            Shape targetShape = model.expectShape(memberShape.getTarget());
+            if (targetShape.hasTrait(StreamingTrait.class)
+                    && targetShape.hasTrait(RequiresLengthTrait.class)) {
+                events.add(error(operationShape, String.format(
+                        "The requestCompression trait can only be applied to operations which do not have "
+                                + "input members that target shapes with both the streaming and requiresLength"
```

Comment:
Missing space at the end.
```suggestion
                                + "input members that target shapes with both the streaming and requiresLength "
```